File: 2024/python/day_2.py
def parse_input(input):
    return [[int(d) for d in line.split()] for line in input.strip().splitlines()]


def sign(i):
    return i / abs(i)


def is_safe(report):
    s = None
    for i in range(len(report) - 1):
        d = report[i] - report[i + 1]
        if not (1 <= abs(d) <= 3):
            return False
        if i == 0:
            s = sign(d)
        elif sign(d) != s:
            return False
    return True


def part1(input):
    return sum(int(is_safe(report)) for report in input)


def problem_dampener(report):
    yield report
    for i in range(len(report)):
        other = list(report)
        other.pop(i)
        yield other


def part2(input):
    return sum(
        int(any(is_safe(r) for r in problem_dampener(report))) for report in input
    )


def main():
    with open("../input/day_2.txt", "r") as f:
        input = parse_input(f.read())

    print("part1:", part1(input))
    print("part2:", part2(input))


if __name__ == "__main__":
    main()


EXAMPLE_1 = """
7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9
"""


def test_part1():
    assert part1(parse_input(EXAMPLE_1)) == 2


def test_part2():
    assert part2(parse_input(EXAMPLE_1)) == 4

File: 2024/golang/day_2/day_2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Input = [][]int

func parseInput(input string) Input {
	var reports [][]int
	sc := bufio.NewScanner(strings.NewReader(input))
	for sc.Scan() {
		fields := strings.Fields(sc.Text())
		report := make([]int, len(fields))
		for i := 0; i < len(fields); i++ {
			v, err := strconv.Atoi(fields[i])
			if err != nil {
				panic(err)
			}
			report[i] = v
		}
		reports = append(reports, report)
	}
	return reports
}

func intAbs(i int) int {
	if i < 0 {
		return -i
	} else {
		return i
	}
}

func intSign(i int) bool {
	if i < 0 {
		return false
	} else {
		return true
	}
}

func isSafe(report []int) bool {
	sign := true
	for i := 0; i < len(report)-1; i++ {
		d := report[i] - report[i+1]
		if intAbs(d) < 1 || intAbs(d) > 3 {
			return false
		}
		if i == 0 {
			sign = intSign(d)
		} else if intSign(d) != sign {
			return false
		}
	}
	return true
}

func part1(input Input) int {
	count := 0
	for _, report := range input {
		if isSafe(report) {
			count += 1
		}
	}
	return count
}

func problemDampener(report []int) [][]int {
	result := make([][]int, len(report))
	for i := 0; i < len(report); i++ {
		other := make([]int, len(report)-1)
		j := 0
		for k := 0; k < len(report); k++ {
			if k != i {
				other[j] = report[k]
				j += 1
			}
		}
		result[i] = other
	}
	return result
}

func part2(input Input) int {
	count := 0
	for _, report := range input {
		if isSafe(report) {
			count += 1
		} else {
			for _, report := range problemDampener(report) {
				if isSafe(report) {
					count += 1
					break
				}
			}
		}
	}
	return count
}

func main() {
	b, err := os.ReadFile("../input/day_2.txt")
	if err != nil {
		panic(err)
	}

	input := parseInput(string(b))
	fmt.Println("part1:", part1(input))
	fmt.Println("part2:", part2(input))
}
